### PR TITLE
[WebNN EP] Support numThreads option for WebNN CPU device

### PR DESCRIPTION
**Pull request by Honry**

### CODE REVIEW

**Comment by Honry** [comment]
@fdwr, @guschmue, PTAL, thanks!

**Comment by guschmue** [comment]
lint is nagging:
image

File: onnxruntime/core/session/provider_registration.cc [hunk 1]
```diff
@@ -104,8 +104,10 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsAppendExecutionProvider,
   } else if (strcmp(provider_name, "WEBNN") == 0) {
 #if defined(USE_WEBNN)
     std::string deviceType = options->value.config_options.GetConfigOrDefault("deviceType", "cpu");
+    std::string numThreads = options->value.config_options.GetConfigOrDefault("numThreads", "0");
```
[the remark this answers is not in the document]
Comment (fdwr) [reply onnxruntime/core/session/provider_registration.cc:107]: Never mind, it does. #Resolved

File: js/web/lib/wasm/session-options.ts [hunk 2]
```diff
@@ -75,6 +75,19 @@ const setExecutionProviders =
                   checkLastError(`Can't set a session config entry: 'deviceType' - ${webnnOptions.deviceType}.`);
                 }
               }
+              if (webnnOptions?.numThreads) {
+                let numThreads = webnnOptions.numThreads;
+                // Just ignore invalid webnnOptions.numThreads.
+                if (typeof numThreads != 'number' || !Number.isInteger(numThreads) || numThreads < 0) {
```
Comment (fdwr) [inline js/web/lib/wasm/session-options.ts:81]: Is `instanceof` better here, maybe avoiding the string compare? (I barely know any JS, but the deeper I dig into it, the messier and less consistent the rules appear 😅)

**Review by fdwr — APPROVED** [review]
When you say "WebNN CPU device", does that mean XNNPack since that's the WebNN CPU backend (because I'm not seeing how this parameter gets forwarded along to WebNN for XNNPack), or is for the CPU EP within ORT when fallback occurs?

File: js/web/lib/wasm/session-options.ts (hunk 2 above)
Comment (Honry) [reply js/web/lib/wasm/session-options.ts:81]: 'instaceof' should not be used for primitive types in JavaScript, e.g. 
```
1 instanceof Number; // false
typeof 1 == 'number'; // true
```

**Comment by Honry** [comment]
> When you say "WebNN CPU device", does that mean XNNPack since that's the WebNN CPU backend

Yeah, right, it means the XNNPack backend. And as you know, `numThreads` only work for WebNN `deviceType=cpu`.

File: js/web/lib/wasm/session-options.ts (hunk 2 above)
Comment (fdwr) [reply js/web/lib/wasm/session-options.ts:81]: > 1 instanceof Number; // false

🤦‍♂️ Oh Javascript...

**Comment by guschmue** [comment]
there is some ut failure. Don't think any of the changes here should do this but its new.
Will look at it.
image

**Comment by fdwr** [comment]
This must be unrelated to this change, given it happens in the TVM pass 🤨:

❌ `Windows_CI / Onnxruntime-TVM (pull_request)`

```
https://github.com/microsoft/onnxruntime/actions/runs/6621336411/job/18111337280?pr=18054#step:9:12453

1: D:\a\onnxruntime\onnxruntime\onnxruntime\test\framework\function_test.cc(585): error: Expected equality of these values:
1:   1
1:   model_proto.functions_size()
1:     Which is: 0
1: Stack trace:
1:   00007FF77725B336: (unknown)
1:   00007FF777283F86: (unknown)
1:   00007FF77728437E: (unknown)
1:   00007FF777284F53: (unknown)
1: ... Google Test internal frames ...
1: 
1: [  FAILED  ] FunctionTest.TestInlinedLocalFunctionNotRemoved (24 ms)
```

**Review by guschmue — APPROVED** [review]

**Comment by Honry** [comment]
@guschmue, what happened in the CI?

**Comment by guschmue** [comment]
the build expired. Let me start a new build.

**Pull request merged**